### PR TITLE
Refactor pkg/subctl/cmd/cloud/* for GCP

### PR DESCRIPTION
**Pull request by Jaanki**
Signed-off-by: Janki Chhatbar <jchhatba@redhat.com>

### CODE REVIEW

File: pkg/cloud/cleanup/gcp.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+package cleanup
+
+import (
+	"github.com/submariner-io/admiral/pkg/reporter"
+	"github.com/submariner-io/cloud-prepare/pkg/api"
+	"github.com/submariner-io/submariner-operator/internal/restconfig"
+	"github.com/submariner-io/submariner-operator/pkg/cloud/gcp"
+)
+
+var gcpConfig gcp.Config
```
Comment (tpantelis) [inline pkg/cloud/cleanup/gcp.go]: Why is this global? It's never populated. It seems this should be passed into `GCP` below.

File: pkg/cloud/cleanup/gcp.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+				return err
+			}
+
+			return cloud.CleanupAfterSubmariner(status)
+		})
+	if err != nil {
+		return status.Error(err, "Failed to cleanup GCP cloud")
+	}
+
+	return nil
```
Comment (tpantelis) [inline pkg/cloud/cleanup/gcp.go]:
```suggestion
	return status.Error(err, "Failed to cleanup GCP cloud")
```

File: pkg/cloud/gcp/gcp.go [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+	Region           string
+	ProjectID        string
+	CredentialsFile  string
+	OcpMetadataFile  string
+	GWInstanceType   string
+}
+
+// RunOn runs the given function on GCP, supplying it with a cloud instance connected to GCP and a reporter that writes to CLI.
+// The functions makes sure that infraID and region are specified, and extracts the credentials from a secret in order to connect to GCP.
+func RunOn(restConfigProducer restconfig.Producer, config *Config, creds *google.Credentials, status reporter.Interface,
```
Comment (tpantelis) [inline pkg/cloud/gcp/gcp.go]:
```suggestion
func RunOn(restConfigProducer *restconfig.Producer, config *Config, creds *google.Credentials, status reporter.Interface,
```

File: pkg/cloud/gcp/gcp.go [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+	if err != nil {
+		return nil, errors.Wrapf(err, "error reading file %q", credentialsFile)
+	}
+
+	creds, err := google.CredentialsFromJSON(context.TODO(), authJSON, dns.CloudPlatformScope)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error parsing credentials file")
+	}
+
+	return creds, nil
```
Comment (tpantelis) [inline pkg/cloud/gcp/gcp.go]:
```suggestion
	return creds, errors.Wrapf(err, "error parsing credentials file")
```

File: pkg/cloud/prepare/gcp.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+	"github.com/submariner-io/admiral/pkg/reporter"
+	"github.com/submariner-io/cloud-prepare/pkg/api"
+	"github.com/submariner-io/submariner-operator/internal/cli"
+	"github.com/submariner-io/submariner-operator/internal/restconfig"
+	"github.com/submariner-io/submariner-operator/pkg/cloud"
+	"github.com/submariner-io/submariner-operator/pkg/cloud/gcp"
+	"golang.org/x/oauth2/google"
+)
+
+func GCP(restConfigProducer *restconfig.Producer, ports cloud.Ports, config *gcp.Config, creds *google.Credentials,
```
Comment (tpantelis) [inline pkg/cloud/prepare/gcp.go]:
```suggestion
func GCP(restConfigProducer *restconfig.Producer, ports *cloud.Ports, config *gcp.Config, creds *google.Credentials,
```

File: pkg/cloud/prepare/gcp.go [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+				}
+			}
+
+			return cloud.PrepareForSubmariner(input, status)
+		})
+	if err != nil {
+		return status.Error(err, "Failed to prepare GCP cloud")
+	}
+
+	return nil
```
Comment (tpantelis) [inline pkg/cloud/prepare/gcp.go]:
```suggestion
	return status.Error(err, "Failed to prepare GCP cloud")
```

File: pkg/subctl/cmd/cloud/cleanup/gcp.go [hunk 7]
```diff
@@ -20,38 +20,31 @@ package cleanup
...
-	"github.com/submariner-io/cloud-prepare/pkg/api"
 	"github.com/submariner-io/submariner-operator/internal/cli"
 	"github.com/submariner-io/submariner-operator/internal/exit"
+	"github.com/submariner-io/submariner-operator/internal/restconfig"
+	"github.com/submariner-io/submariner-operator/pkg/cloud/cleanup"
+	gcpcloud "github.com/submariner-io/submariner-operator/pkg/cloud/gcp"
 	"github.com/submariner-io/submariner-operator/pkg/subctl/cmd/cloud/gcp"
 )
 
+var gcpConfig gcpcloud.Config
```
Comment (tpantelis) [inline pkg/subctl/cmd/cloud/cleanup/gcp.go]: This is unused and can be removed.

File: pkg/subctl/cmd/cloud/prepare/gcp.go [hunk 8]
```diff
@@ -20,64 +20,52 @@ package prepare
...
+	cloudgcp "github.com/submariner-io/submariner-operator/pkg/cloud/gcp"
+	"github.com/submariner-io/submariner-operator/pkg/cloud/prepare"
 	"github.com/submariner-io/submariner-operator/pkg/subctl/cmd/cloud/gcp"
 )
 
+var gcpConfig cloudgcp.Config
+
 // NewCommand returns a new cobra.Command used to prepare a cloud infrastructure.
-func newGCPPrepareCommand() *cobra.Command {
+func newGCPPrepareCommand(restConfigProducer *restconfig.Producer, ports cloud.Ports) *cobra.Command {
```
Comment (tpantelis) [inline pkg/subctl/cmd/cloud/prepare/gcp.go]:
```suggestion
func newGCPPrepareCommand(restConfigProducer *restconfig.Producer, ports *cloud.Ports) *cobra.Command {
```

**Review by tpantelis — CHANGES_REQUESTED** [review]

File: pkg/subctl/cmd/cloud/cleanup/gcp.go (hunk 7 above)
Comment (Jaanki) [reply pkg/subctl/cmd/cloud/cleanup/gcp.go]: Unused yes, but `AddGCPFlags()` needs it as it is needed for `prepare`. I made the variable local instead of global.

File: pkg/cloud/gcp/gcp.go (hunk 3 above)
Comment (Jaanki) [reply pkg/cloud/gcp/gcp.go]: I tried to keep the original function parameters similar.

File: pkg/subctl/cmd/cloud/cleanup/gcp.go [hunk 9]
```diff
@@ -20,38 +20,31 @@ package cleanup
...
+	"github.com/submariner-io/submariner-operator/internal/restconfig"
+	"github.com/submariner-io/submariner-operator/pkg/cloud/cleanup"
+	gcpcloud "github.com/submariner-io/submariner-operator/pkg/cloud/gcp"
 	"github.com/submariner-io/submariner-operator/pkg/subctl/cmd/cloud/gcp"
 )
 
 // newGCPCleanupCommand returns a new cobra.Command used to prepare a cloud infrastructure.
-func newGCPCleanupCommand() *cobra.Command {
+func newGCPCleanupCommand(restConfigProducer restconfig.Producer) *cobra.Command {
+	var gcpConfig gcpcloud.Config
```
Comment (tpantelis) [inline pkg/subctl/cmd/cloud/cleanup/gcp.go]: This is still unused for cleanup so L47 below has no actual affect.

**Review by tpantelis — CHANGES_REQUESTED** [review]

File: pkg/subctl/cmd/cloud/utils/utils.go [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+	RegionFlag    = "region"
+	ProjectIDFlag = "project-id"
+)
+
+// ExpectFlag exits with an error if the flag value is empty.
+func ExpectFlag(flag, value string) {
+	if value == "" {
+		exit.WithMessage(fmt.Sprintf("You must specify the %v flag", flag))
+	}
+}
```
Comment (tpantelis) [inline pkg/subctl/cmd/cloud/utils/utils.go]: See [comment](https://github.com/submariner-io/submariner-operator/pull/2048/files#r866386963)

File: pkg/subctl/cmd/cloud/cleanup/gcp.go [hunk 11]
```diff
@@ -20,38 +20,50 @@ package cleanup
...
+			} else {
+				utils.ExpectFlag(utils.InfraIDFlag, gcpConfig.InfraID)
+				utils.ExpectFlag(utils.RegionFlag, gcpConfig.Region)
+				utils.ExpectFlag(utils.ProjectIDFlag, gcpConfig.Region)
+			}
 
-	return cmd
-}
+			gcpConfig.GWInstanceType = ""
+			gcpConfig.DedicatedGateway = false
```
Comment (tpantelis) [inline pkg/subctl/cmd/cloud/cleanup/gcp.go:54]: Why is this necessary?

**Review by tpantelis — CHANGES_REQUESTED** [review]

File: pkg/subctl/cmd/cloud/cleanup/gcp.go (hunk 11 above)
Comment (Jaanki) [reply pkg/subctl/cmd/cloud/cleanup/gcp.go:54]: Well, `cleanup` passes empty `GW instance type` and false `dedicated GWs` as arguments to `gcp,RunOn` (along with infra-id, region, etc...). I understand that the default values for string is empty and bool is false. I just mentioned that explicitly here.

**Review by tpantelis — APPROVED** [review]

**Comment by dfarrell07** [comment]
Was https://github.com/submariner-io/submariner-operator/pull/2041/commits/0121836fb7afc35fb203294d6f7779f032999da7 reviewed by @tpantelis? It seems to be newer than his review.

**Comment by dfarrell07** [comment]
> enabled auto-merge (squash)

That would destroy the commit history preserved by these mv/etc commits

**Comment by tpantelis** [comment]
> > enabled auto-merge (squash)
> 
> That would destroy the commit history preserved by these mv/etc commits

yup which is what happened with https://github.com/submariner-io/submariner-operator/pull/2048 😦 I'll revert it.

**Comment by dfarrell07** [comment]
> Was [0121836](https://github.com/submariner-io/submariner-operator/commit/0121836fb7afc35fb203294d6f7779f032999da7) reviewed by @tpantelis? It seems to be newer than his review.

This would be the ideal time for !fixup commits BTW, (I assume) a tweak to a commit after a review of a complex patchset.

**Comment by Jaanki** [comment]
The last commit addresses just the linting issues.

**Review by dfarrell07 — APPROVED** [review]
